dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs

```diff
@@ -0,0 +1,274 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace Microsoft.SemanticKernel;
+
+/// <summary>
+/// Represents a step in a process that is running in-process.
+/// </summary>
+internal class LocalStep : KernelProcessMessageChannel
+{
+    /// <summary>
+    /// The generic state type for a process step.
+    /// </summary>
+    private static readonly Type s_genericType = typeof(KernelProcessStep<>);
+    private Dictionary<string, Dictionary<string, object?>?>? _inputs;
+    private Dictionary<string, Dictionary<string, object?>?>? _initialInputs;
+    private ILogger? _logger;
+
+    private readonly Dictionary<string, KernelFunction> _functions;
+    private readonly Kernel _kernel;
+    private readonly Queue<KernelProcessEvent> _eventQueue;
+
+    protected readonly string? ParentProcessId;
+    protected readonly ILoggerFactory? LoggerFactory;
+    protected Dictionary<string, List<KernelProcessEdge>>? _outputEdges;
+
+    public readonly string Name;
+    public readonly string Id;
+
+    public LocalStep(string name, string id, Kernel kernel, string? parentProcessId = null, ILoggerFactory? loggerFactory = null)
+    {
+        Verify.NotNullOrWhiteSpace(name);
+        Verify.NotNullOrWhiteSpace(id);
+        Verify.NotNull(kernel);
+
+        this.Name = name;
+        this.Id = id;
+        this.ParentProcessId = parentProcessId;
+        this.LoggerFactory = loggerFactory;
+
+        this._kernel = kernel;
+        this._eventQueue = new();
+        this._functions = [];
+    }
+
+    public virtual async ValueTask InitializeAsync(KernelProcessStepInfo stepInfo)
+    {
+        this._logger = this.LoggerFactory?.CreateLogger(stepInfo.InnerStepType) ?? new NullLogger<LocalStep>();
+        this._outputEdges = stepInfo.Edges.ToDictionary(kvp => kvp.Key, kvp => kvp.Value.ToList());
+
+        // Instantiate an instance of the inner step object
+        KernelProcessStep stepInstance = (KernelProcessStep)ActivatorUtilities.CreateInstance(this._kernel.Services, stepInfo.InnerStepType);
+        var kernelPlugin = KernelPluginFactory.CreateFromObject(stepInstance);
+
+        // Load the kernel functions
+        foreach (KernelFunction f in kernelPlugin)
+        {
+            Verify.NotNull(f);
+            var cloned = f.Clone(stepInfo.State.Name!);
+            this._functions.Add(cloned.Name, cloned);
```

Comment:
Why is the cloning needed?

Reply:
Good catch, this is not needed anymore. Updated it to set the plugin name when loading the plugin from the factory.